README.md
# Node Replay


### When API testing slows you down: record and replay HTTP responses like a boss

Things that will ruin your day when tests make HTTP requests to other services:

- That other service has the uptime of Twitter's API
- Network late ............ ncy
- Being-rate limited and having to wait an hour for the next test run
- Same request returns different result each time
- Everyone else on the network is deep in BitTorrent territory

Things **node-replay** can do to make these problems go away:

- Record API response once, replay as often as necessary
- Stub HTTP requests (TBD)
- Replay different responses to same request (great for testing error handling)
- Not suck


## How to use node-replay

Like this:

```bash
npm install replay
```

Now write some simple test case:

```javascript
const assert  = require('assert');
const HTTP    = require('http');
const Replay  = require('replay');

HTTP.get({ hostname: 'www.iheartquotes.com', path: '/api/v1/random' }, function(response) {
  var body = '';
  response.on('data', function(chunk) {
    response.body = response.body + chunk;
  });
  response.on('end', function() {

    // Now check the request we made to the I <3 Quotes API
    assert.equal(response.statusCode, 200);
    assert.equal(response.body, 'Oxymoron 2. Exact estimate\n\n[codehappy] http://iheartquotes.com/fortune/show/38021\n');
    console.log('Woot!');

  });
});
```

This, of course, will fail the first time you run it.  You'll see:

```
Error: Connection to http://www.iheartquotes.com:80/api/v1/random refused: not recording and no network access
    at Array.0 (/Users/assaf/projects/node-replay/lib/replay/proxy.coffee:87:21)
    at EventEmitter._tickCallback (node.js:192:40)
```

Unless you tell it otherwise, **node-replay** runs in `replay` mode.  In this
mode it will replay any previously captured HTTP response, but it will not allow
any outgoing network connection.

That's the default mode for running tests.  "Why?" you ask.  Good question.
Running in `replay` mode forces any test you run to use recorded reponses, and
so it will run (and fail or pass) the same way for anyone else, any other day of
the week, on whatever hardware they use.  Even if they're on the AT&T network.

Running in `replay` mode helps you write repeatable tests.  Repeatable tests are
a Good Thing.

So the first thing you want to do to get that test to pass, is to run
**node-replay** in `record` mode.  In this mode it will replay any recorded
response, but if no response was recorded, it will make a request to the server
and capture the response.

Let's do that:

```bash
REPLAY=record node test.js
```

That wasn't too hard, but the test is still failing.  "How?"  You must be
wondering and scratching your head in total disbelief.  It's actually quite
simple.

Every request you make to 'I 3> Quotes' returns a different quote, and that test
is looking for a very specific quote.  So the test will fail, and each time fail
with a different error.

So one way we can fix this test is to change the assertion.  Look at the error
message, get the actual quote and make the assertion look for that value.

Now run the test:

```bash
$ node test.js
=> Woot!
```

Did the test pass?  Of course it did.  Run it again.  Still passing?  Why, yes.

So let's have a look at that captured response.  All the respones recorded for
'I <3 Quotes>' will be listed here:

```bash
ls fixtures/www.iheartquotes.com/
```

There should be only one file there, since we only recorded one response.  The
file name is a timestamp, but feel free to rename it to something more
descriptive.

The name of a response file doesn't matter, it can be whatever you want.  The
name of the directory does, though, it matches the service hostname (and port
when not 80).

So that was one way to fix the failing test.  Another one is to change the
recorded response to match the assertion.  Being able to edit (and create new)
responses is quite important.  Sometimes it's the easiest way to create mock
responses for testing, e.g. if you're trying to test failure conditions that are
hard to come by.

So let's edit the response file and change the body part, so the entire response
reads like this:

```
/api/v1/random

HTTP/1.1 200 OK
server: nginx/0.7.67
date: Fri, 02 Dec 2011 02:58:03 GMT
content-type: text/plain
connection: keep-alive
etag: "a7131ebc1e81e43ea9ecf36fa2fdf610"
x-ua-compatible: IE=Edge,chrome=1
x-runtime: 0.158080
cache-control: max-age=0, private, must-revalidate
content-length: 234
x-varnish: 2274830138
age: 0
via: 1.1 varnish

Oxymoron 2. Exact estimate

[codehappy] http://iheartquotes.com/fortune/show/38021
```

All responses are stored as text files using the simplest format ever, so you
can edit them in Vim, or any of the many non-Vim text editors in existence:

- First comes the request path (including query string)
- Followed by any headers sent as part of the request (like `Accept` and `Authorization`)
- Then an empty line
- Next the response status code and (optional) HTTP version number
- Followed by any headers sent as part of the response
- Then another empty line
- And the rest taken by the response body

If you need to use regular expressions to match the request URL, add `REGEXP`
between the method and path, for example:

```
GET REGEXP /\/Aregexp\d/i

HTTP/1.1 200 OK
Content-Type: text/html
```


## Settings

We've got them.  Just enough to make you happy and not enough to take all day to
explain.

The first and most obvious is the mode you run **node-replay** in:

**bloody** -- All requests go out, none get replayed.  Use this if you want to
remember what life was before you started using **node-replay**.  Also, to test
your code against changes to 3rd party API, because these do happen.  Too often.

**cheat** -- Replays recorded responses, and allow HTTP outbound requests.  This
is mighty convenient when you're writing new tests or changing code to make new,
un-recorded HTTP requests, but you haven't quite settled on which requests to
make, so you don't want any responses recorded quite yet.

**record** -- Replays recorded responses, or captures responses for future
replay.  Use this whenever you're writing new tests or code that makes new HTTP
requests.

**replay** -- Replays recorded responses, does not allow outbound requests.
This is the default mode.  That's another way of saying, "you'll be running in
this mode most of the time".

You can set the mode by setting the environment variable `REPLAY` to one of
these values:

```bash
REPLAY=record node test.js
```

Of from your code by setting `replay.mode`:

```javascript
const Replay = require('replay');
Replay.mode = 'record';
```

Of course, **node-replay** needs to store all those captured responses somewhere,
and by default it will put them in the directory `fixtures`.  Bet you have an
idea for a better directory name.  Easy to change.

Like this:

```javascript
Replay.fixtures = __dirname + '/fixtures/replay';
```

You can tell **node-replay** what hosts to treat as "localhost".  Requests to
these hosts will be routed to 127.0.0.1, without capturing or replay.  This is
particularly useful if you're making request to a test server and want to use
the same URL as production.

For example:

```javascript
Replay.localhost('www.example.com');
```

If you don't want requests going to specific server, you can add them to the
drop list.  For example, Google Analytics, where you don't care that the request
go through, and you don't want to record it.

```javascript
Replay.drop('www.google-analytics.com', 'rollbar.com');
```

Likewise, you can tell **node-replay** to pass through requests to specific hosts:

```javascript
Replay.passThrough('s3.amazonaws.com');
```

If you're running into trouble, try turning debugging mode on.  It helps.
Sometimes.

```bash
$ DEBUG=replay node test.js
=> Requesting http://www.iheartquotes.com:80/api/v1/random
=> Woot!
```

By default, **node-replay** will record the following headers with each request,
and use only these headers when matching pre-recorded requests:

- Headers starting with `Accept` (eg `Accept-Encoding`)
- `Authorization`
- `Body` (used to match the body of a POST request)
- `Content-Type`
- `Host`
- Headers starting with `If-` (eg `If-Modified-Since`)
- Headers starting with `X-` (eg `X-Requested-With`)

You can modify the list of matched headers, adding or removing headers, by
changing the value of `Replay.headers`.  The value is an array of regular
expressions.

For example, to capture `content-length` (useful with file uploads):

```javascript
Replay.headers.push(/^content-length/);
```

Since headers are case insensitive, we always match on the lower case name.

If you want more control over the responses that you record (for example weeding out error responses for 
really flaky backends or when you are polling waiting for something to happen), you can use recordReponseControl 
to excercise fine grain control over the recording process. For example, to only save the responses from myhostame.com port 8080 if the response was successful:

```javascript
Replay.recordResponseControl = {
  "myhostname.com:8080" : function(request, response) {
    return response.statusCode < 400;
  }
};
```


## Geeking

To make all that magic possible, **node-replay** replaces
`require('http').request` with its own method.  That method returns a
`ProxyRequest` object that captures the request URL, headers and body.

When it's time to fire the request, it gets sent through a chain of proxies.
The first proxy to have a response, returns it (via callback, this is Node.js
after all).  That terminates the chain.  A proxy that doesn't have a response
still has to call the callback, but with no arguments.  The request will then
pass to the next proxy down the chain.

The proxy chain looks something like this:

- Logger dumps the request URL when running with `DEBUG=replay`
- The pass-through proxy will pass the request directly to the server in `bloody` mode, or when talking to `localhost`
- The recorder proxy will either replay a captured request (if it has one), talk to the server and capture the response
  (in `record` mode), or pass to the next proxy
- The pass-through proxy (2nd one) will pass the request to the server in `cheat` mode, return nothing in all other
  modes

Loading pre-recorded responses to memory, from where they can be replayed, and
storing new ones on disk, is handled by ... cue big band ... the `Catalog`.


## Final words

**node-replay** is released under the MIT license.  Pull requests are welcome.



CHANGELOG.md
## Version 2.4.0 2018-11-30

Actually fix Node 10.

Minor bump because we changed the way we patch the HTTP/HTTPS modules.


## Version 2.3.0 2018-04-29

Tested with Node 10.

add support for matching body with regular expression #145


## Version 2.2.0 2018-04-02

Remove Content-Length after unzipping data #143

Ignore dot files #140

Fix for request library SSL authorization check #135


## Version 2.1.4 2017-10-19

ADDED control over which responses are recorded #68

  Replay.recordResponseControl = {
    "myhostname.com:8080" : function(request, response) {
      return response.statusCode < 400;
    }
  };

ADDED record unzipped reply #72

ADDED support for passing-through client cert and key #99

ADDED backwards-compatible support for old response status formatting #132


## Version 2.1.3 2017-10-17

FIXED for request library SSL authorization check

Updated dependencies


## Version 2.1.2 2016-09-13

FIXED localhosts not passed through


## Version 2.1.1 2016-09-12

FIXED compatibility with Node LTS and Current


## Version 2.1.0 2015-05-03

CHANGED rewrote in ES6


## Version 2.0.6 2015-04-30

FIXED pass through when sending request bodies


## Version 2.0.5 2015-04-28

FIXED SSL requests when not providing an explicit `protocol` option.


## Version 2.0.4 2015-03-10

FIXED problem with https support on 0.10


## Version 2.0.3 2015-03-10

ADDED Supports streaming API v2 (0.12/iojs)

ADDED Supports new additions to HTTP API (0.12/iojs)

CHANGED Replay.allow is now Replay.passThrough, a better name since the behavior
is to pass through requests directly to the target server.

CHANGED Replay.ignore is now Replay.drop, a better name since the behavior is to
drop the connection (you'll get ECONNREFUSED).

CHANGED New way for writing HTTP response line:

  HTTP/1.1 200 OK
  HTTP/1.1 404 Not Found

CHANGED To enable debugging, run with environment variable DEBUG=replay

FIXED HTTPS requests looping forvever


## Version 1.12.0 2014-12-04

ADDED you can now allow, ignore and localhost multiple domains

```
Replay.localhos('*.example.com');
```


## Version 1.11.0 2014-11-10

ADDED now saving query string as part of request URL

FIXED when ignoring request, throw ECONNREFUSED error

FIXED handle case where header value is zero (number)


## Version 1.10.3 2014-04-19

ADDED 127.0.0.1 to default localhost addresses (Rajit Singh)

FIXED double callback issue on catalog save (Itay Adler)


## Version 1.10.2 2014-04-19

FIXED bug with header stringifying (for realz now)


## Version 1.10.1 2014-04-19

FIXED bug with header stringifying

CHANGED run with DEBUG=replay or DEBUG=all


## Version 1.10.0 2014-04-18

CHANGED Support recording and serving images (Rajit).

CHANGED Support dynamic switching of fixtures (Pål Ruud):

  Replay.fixtures = "fixture-directory-2"

FIXED compatibility with superagent and supertest (Jerome Touffe-Blin)

FIXED support http.get([string])

FIXED request headers are stringified before processing


## Version 1.9.1 2013-12-05

Fix "Recursive nextTick" warning (Drew Stokes)


## Version 1.9.0 2013-11-04

Added support for specifying which headers are stored when recording, and used
for matching when replaying.  Edit the list of regular expressions in
`Replay.headers`.

Fixed a bug in capturing and filtering on POST request body.


## Version 1.8.0 2013-10-01

Added support for capturing and filtering on POST request body (Jerome
Gravel-Niquet)


## Version 1.7.0 2013-05-06

Support Windows by creating directories of the form host-port instead of
host:port.

Updated dev dependencies, testing with Express 3.2.


## Version 1.6.2 2013-03-19

Add randomness to uid to deal with fast requests and timekeeper (Nate Murray).

Add support for HTTPS (Nate Murray).

Tested with Node 0.10.0.


## Version 1.6.1 2013-02-18

Improved logging options:
- Replay will now emit errors to the console unless otherwise told
- You can tell Replay to not emit errors to console by setting
  Replay.silent = true
- You can also tell Replay to emit errors an logs messages elsewhere by setting
  Replay.logger (defaults to console)
- You can listen to all events via Replay.on("error", function(error) { })

No deprecation messages on Node 0.8.

Do not assume /tmp directory exists.

Works with request 2.11 by implementing a do nothing response.connection event


## Version 1.5.3 2012-06-30

Removed engine dependency. Works on 0.8.1 now.


## Version 1.5.2 2012-05-15

Do not fail on headers with empty value.


## Version 1.5.1 2012-05-14

When matching request against headers, also match the Authorization header
(David John).


## Version 1.5.0 2012-05-08

Properly handle repeating headers (e.g. set-cookie) by storing and reading
multiple entries.


## Version 1.4.4 2012-05-02

Filter out request headers *not* response headers.


## Version 1.4.3 2012-05-02

Precompile before publishing, no longer requires Coffee-Script to run.


## Version 1.4.2 2012-05-02

Added support for HTTPS (Jerome Gravel-Niquet)


## Version 1.4.1 2012-04-30

Do not store request headers we don't care for.


## Version 1.4.0 2012-04-30

Replay files can now use REGEXP to match request URL (Jerome Gravel-Niquet)


## Version 1.3.1 2012-03-15

Accept replay documents with nothing but method and path.


## Version 1.3.0 2012-03-15

Fix status code being string instead of integer.

Fix handling of fixtures with empty body.


## Version 1.2.3 2012-01-16

Support (or don't fail on) Web Sockets.

Fix non-working `Replay.localhost`.


## Version 1.2.2 2011-12-27

There may be hosts you don't care to record/replay: it doesn't matter if requests to these hosts succeed or not, and you
don't care to manage their recorded file.  You can just add those to the ignore list:

    Replay.ignore "www.google-analytics.com", "airbrake.io"

The `allow`, `ignore` and `localhost` methods now accept multiple arguments.


## Version 1.2.1 2011-12-27

Bug fix to DNS hack.


## Version 1.2.0 2011-12-27


src/proxy.js
// A proxy is a function that receives two arguments, a request object and a callback.
//
// If it can generate a respone, it calls callback with null and the response object.  Otherwise, either calls callback
// with no arguments, or with an error to stop the processing chain.
//
// The request consists of:
// url     - URL object
// method  - Request method (lower case)
// headers - Headers object (names are lower case)
// body    - Request body, an array of body part/encoding pairs
//
// The response consists of:
// version   - HTTP version
// status    - Status code
// headers   - Headers object (names are lower case)
// body      - Array of body parts
// trailers  - Trailers object (names are lower case)
//
// This file defines ProxyRequest, which acts as an HTTP ClientRequest that captures the request and passes it to the
// proxy chain, and ProxyResponse, which acts as an HTTP ClientResponse, playing back a response it received from the
// proxy.
//
// No actual proxies defined here.


const assert            = require('assert');
const { EventEmitter }  = require('events');
const HTTP              = require('http');
const HTTPS             = require('https');
const Stream            = require('stream');
const URL               = require('url');


// HTTP client request that captures the request and sends it down the processing chain.
module.exports = class ProxyRequest extends HTTP.IncomingMessage {

  constructor(options = {}, proxy) {
    super();
    this.proxy          = proxy;
    this.method         = (options.method || 'GET').toUpperCase();
    const protocol      = options.protocol || (options._defaultAgent && options._defaultAgent.protocol) || 'http:';
    const [host, port]  = (options.host || options.hostname).split(':');
    const realPort      = options.port || port || (protocol === 'https:' ? 443 : 80);
    this.url            = URL.parse(`${protocol}//${host || 'localhost'}:${realPort}${options.path || '/'}`, true);
    this.path           = this.url.path;
    this.auth           = options.auth;
    this.agent          = options.agent || (protocol === 'https:' ? HTTPS.globalAgent : HTTP.globalAgent);
    this.cert           = options.cert;
    this.key            = options.key;
    this.headers        = {};
    if (options.headers)
      for (let name in options.headers) {
        let value = options.headers[name];
        if (value != null)
          this.headers[name.toLowerCase()] = value.toString();
      }
  }

  flushHeaders() {
  }

  setHeader(name, value) {
    assert(!this.ended, 'Already called end');
    assert(!this.body, 'Already wrote body parts');
    this.headers[name.toLowerCase()] = value;
  }

  getHeader(name) {
    return this.headers[name.toLowerCase()];
  }

  removeHeader(name) {
    assert(!this.ended, 'Already called end');
    assert(!this.body, 'Already wrote body parts');
    delete this.headers[name.toLowerCase()];
  }

  addTrailers(trailers) {
    this.trailers = trailers;
  }

  setTimeout(timeout, callback) {
    if (callback)
      setImmediate(callback);
  }

  setNoDelay(/*nodelay = true*/) {
  }

  setSocketKeepAlive(/*enable = false, initial*/) {
  }

  write(chunk, encoding, callback) {
    assert(!this.ended, 'Already called end');
    this.body = this.body || [];
    this.body.push([chunk, encoding]);
    if (callback)
      setImmediate(callback);
  }

  end(data, encoding, callback) {
    assert(!this.ended, 'Already called end');

    if (typeof data === 'function')
      [ callback, data ] = [ data, null ];
    else if (typeof encoding === 'function')
      [ callback, encoding ] = [ encoding, null ];

    if (data) {
      this.body = this.body || [];
      this.body.push([ data, encoding ]);
    }
    this.ended = true;

    if (callback)
      setImmediate(callback);

    this.proxy(this, (error, captured)=> {
      // We're not asynchronous, but clients expect us to callback later on
      setImmediate(()=> {
        if (error)
          this.emit('error', error);
        else if (captured) {
          const response = new ProxyResponse(captured);
          this.emit('response', response);
          response.resume();
        } else {
          const error = new Error(`${this.method} ${URL.format(this.url)} refused: not recording and no network access`);
          error.code  = 'ECONNREFUSED';
          error.errno = 'ECONNREFUSED';
          this.emit('error', error);
        }
      });
    });
  }

  flush() {
  }

  abort() {
  }

};


// HTTP client response that plays back a captured response.
class ProxyResponse extends Stream.Readable {

  constructor(captured) {
    super();
    this.once('end', ()=> {
      this.emit('close');
    });

    this.httpVersion      = captured.version || '1.1';
    this.httpVersionMajor = this.httpVersion.split('.')[0];
    this.httpVersionMinor = this.httpVersion.split('.')[1];
    this.statusCode       = parseInt(captured.statusCode || 200, 10);
    this.statusMessage    = captured.statusMessage || HTTP.STATUS_CODES[this.statusCode] || '';
    this.headers          = Object.assign({ }, captured.headers);
    this.rawHeaders       = (captured.rawHeaders || [].slice(0));
    this.trailers         = Object.assign({ }, captured.trailers);
    this.rawTrailers      = (captured.rawTrailers || []).slice(0);
    // Not a documented property, but request seems to use this to look for HTTP parsing errors
    this.connection       = new EventEmitter();
    this._body            = captured.body.slice(0);
    this.client           = { authorized: true }
  }

  _read() {
    const part = this._body.shift();
    if (part)
      this.push(part[0], part[1]);
    else
      this.push(null);
  }

  setTimeout(msec, callback) {
    if (callback)
      setImmediate(callback);
  }

  static notFound(url) {
    return new ProxyResponse({
      status: 404,
      body:   [ `No recorded request/response that matches ${URL.format(url)}` ]
    });
  }

}

